### PR TITLE
feat(do-rpc): Adding a way to instrument DO that use RPC methods

### PR DESCRIPTION
**Pull request by sathiya-s1**
Fixes # [insert GH issue number(s)].

**What this PR solves / how to test:**

### CODE REVIEW

File: src/instrumentation/do-rpc.ts [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+		construct(target, [orig_state, orig_env]: ConstructorParameters<DOClass>) {
+			const trigger: DOConstructorTrigger = {
+				id: orig_state.id.toString(),
+				name: orig_state.id.name,
+			}
+			const constructorConfig = initialiser(orig_env, trigger)
+			const context = setConfig(constructorConfig)
+			const env = instrumentEnv(orig_env)
+			const createDO = () => {
+				orig_state.storage.sql = instrumentSql(orig_state.storage.sql)
```
Comment (sathiya-s1) [inline src/instrumentation/do-rpc.ts:245]: DurableObject constructor asserts that the state is a DurableObjectState, so passing a ProxyHandler doesn't work. Thus this hack.

File: dist/index.js [hunk 2]
```diff
@@ -1786,6 +1789,7 @@ function instrumentD1Fn(fn, dbName, operation) {
 function instrumentD1(database, dbName) {
   const dbHandler = {
     get: (target, prop, receiver) => {
+      console.log("D1 get", prop);
```
Comment (romeupalos) [inline dist/index.js:1792]: Probably remove this before merging.

File: src/instrumentation/do-rpc.ts [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+		attributes[SemanticAttributes.DB_STATEMENT] = sql?.replace(/\n/g, ' ').replace(/\s+/g, ' ').trim();
+	}
+	return {
+		kind: SpanKind.CLIENT,
+		attributes,
+	}
+}
+
+function metaAttributes(meta: SqlStorageIngestResult): Attributes {
+	return {
```
Comment (romeupalos) [inline src/instrumentation/do-rpc.ts:58]: Ideally, you can return the entire `meta` here. Just converting camelCase to snake case.

File: src/instrumentation/do-rpc.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+	}
+	return wrap(fn, fnHandler)
+}
+
+export function instrumentSql(sql: SqlStorage) {
+	const sqlHandler: ProxyHandler<SqlStorage> = {
+		get(target, prop, receiver) {
+			const result = Reflect.get(target, prop, receiver)
+			 if (prop === 'exec') {
```
Comment (romeupalos) [inline src/instrumentation/do-rpc.ts:101]: There are a few more interesting props. In foundation we have a list of them. e.g., `run`, etc.

**Review by romeupalos — APPROVED** [review]

File: src/instrumentation/do-rpc.ts (hunk 4 above)
Comment (sathiya-s1) [reply src/instrumentation/do-rpc.ts:101]: I'm only handling SQL-Lite methods in this PR

https://developers.cloudflare.com/durable-objects/api/storage-api/

File: src/instrumentation/do-rpc.ts (hunk 3 above)
Comment (sathiya-s1) [reply src/instrumentation/do-rpc.ts:58]: The return value (result) is a Cursor. The keys aren't iterable.

File: dist/index.js (hunk 2 above)
Comment (sathiya-s1) [reply dist/index.js:1792]: Done

**Pull request merged**